src/GafferScene/TransformQuery.cpp

```diff
@@ -0,0 +1,522 @@
+    TransformQuery::rotateOrder
+    ()
+    const
+    {
+        int const value = rotateOrderPlug()->getValue();
+        assert( 0 <= value && value < _RotateOrderCount_ );
+        return (RotateOrder) value;
+    }
+
+    void 
+    TransformQuery::affects
+    (
+        Gaffer::Plug const* const ip_input, 
+        AffectedPlugsContainer& io_outputs 
+    )
+    const
+    {
+        ComputeNode::affects( ip_input, io_outputs );
+
+        if( ip_input == outMatrixPlug() )
+        {
+            io_outputs.push_back( outTranslatePlug()->getChild( 0 ) );
+            io_outputs.push_back( outTranslatePlug()->getChild( 1 ) );
+            io_outputs.push_back( outTranslatePlug()->getChild( 2 ) );
+            io_outputs.push_back( outRotatePlug()->getChild( 0 ) );
+            io_outputs.push_back( outRotatePlug()->getChild( 1 ) );
+            io_outputs.push_back( outRotatePlug()->getChild( 2 ) );
+            io_outputs.push_back( outScalePlug()->getChild( 0 ) );
+            io_outputs.push_back( outScalePlug()->getChild( 1 ) );
+            io_outputs.push_back( outScalePlug()->getChild( 2 ) );
+        }
+        else
+        if( ( ip_input == outMatrixPlug() ) ||
+            ( ip_input == spacePlug() ) ||
+            ( ip_input == invertPlug() ) ||
+            ( ip_input == locationPlug() ) ||
+            ( ip_input == relativeLocationPlug() ) ||
+            ( ip_input == scenePlug()->existsPlug() ) ||
+            ( ip_input == scenePlug()->transformPlug() ) )
+        {
+            io_outputs.push_back( outMatrixPlug() );
+        }
+        else
+        if( ip_input == rotateOrderPlug() )
+        {
+            io_outputs.push_back( outRotatePlug()->getChild( 0 ) );
+            io_outputs.push_back( outRotatePlug()->getChild( 1 ) );
+            io_outputs.push_back( outRotatePlug()->getChild( 2 ) );
+        }
+    }
+
+    void
+    TransformQuery::hash
+    (
+        Gaffer::ValuePlug const* const ip_plug,
+        Gaffer::Context const* const ip_context,
+        IECore::MurmurHash& io_hash
+    )
+    const
+    {
+        if( ip_plug == outMatrixPlug() )
+        {
+            ComputeNode::hash( ip_plug, ip_context, io_hash );
+
+            std::string const loc = locationPlug()->getValue();
+            if( ! loc.empty() )
+            {
+                ScenePlug const* const p_scene_plug = scenePlug();
+                assert( p_scene_plug != 0 );
+
+                ScenePlug::ScenePath const path = ScenePlug::stringToPath( loc );
+
+                if( p_scene_plug->exists( path ) )
+                {
+                    switch( space() )
+                    {
+                        case SpaceLocal:
+                        {
+                            io_hash = p_scene_plug->transformHash( path );
+
+                            if( invertPlug()->getValue() )
+                            {
+                                invertPlug()->hash( io_hash );
+                            }
```

Comment:
I think there's quite a subtle bug here. It's completely legitimate to do `io_hash = someInputHash` where compute will return the exact input value corresponding to that hash. But we can't then _append_ to that hash as we do if `invert == true`. The hash we've "borrowed" belongs to another node, and will be initialised with that node's type and the name of the output plug on that node. If we append to it, we can clash with other legitimate hashes for that node that have nothing to do with us. The rule is _either_ call the base class and then append _or_ pass through an input hash, but we can't mix the two.

In this case, I think it might be reasonable to decide that the benefit of passing through the hash doesn't warrant the extra code complexity. Or we'd have to do something like this :

```
if( !invert )
{
   hash = scenePlug()->transformHash( path );
}
else
{
    ComputeNode::hash( plug, context, hash );
    hash.append( scenePlug()->transformHash( path ) );
}
```